// react-native-ui-datepicker/lib/commonjs/components/Calendar.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.default = void 0;
var _react = _interopRequireDefault(require("react"));
var _reactNative = require("react-native");
var _CalendarContext = require("../CalendarContext");
var _Header = _interopRequireDefault(require("./Header"));
var _YearSelector = _interopRequireDefault(require("./YearSelector"));
var _MonthSelector = _interopRequireDefault(require("./MonthSelector"));
var _DaySelector = _interopRequireDefault(require("./DaySelector"));
var _TimeSelector = _interopRequireDefault(require("./TimeSelector"));
var _enums = require("../enums");
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const CalendarView = {
  year: /*#__PURE__*/_react.default.createElement(_YearSelector.default, null),
  month: /*#__PURE__*/_react.default.createElement(_MonthSelector.default, null),
  day: /*#__PURE__*/_react.default.createElement(_DaySelector.default, null),
  time: /*#__PURE__*/_react.default.createElement(_TimeSelector.default, null)
};
const Calendar = ({
  buttonPrevIcon,
  buttonNextIcon
}) => {
  const {
    calendarView,
    mode
  } = (0, _CalendarContext.useCalendarContext)();
  return /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.container,
    testID: "calendar"
  }, mode !== 'time' ? /*#__PURE__*/_react.default.createElement(_Header.default, {
    buttonPrevIcon: buttonPrevIcon,
    buttonNextIcon: buttonNextIcon
  }) : null, /*#__PURE__*/_react.default.createElement(_reactNative.View, {
    style: styles.calendarContainer
  }, CalendarView[calendarView]));
};
const styles = _reactNative.StyleSheet.create({
  container: {
    width: '100%'
  },
  calendarContainer: {
    height: _enums.CALENDAR_HEIGHT,
    alignItems: 'center'
  }
});
var _default = exports.default = Calendar;
