// react-native-ui-datepicker/lib/commonjs/utils.js
"use strict";

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.YEAR_PAGE_SIZE = exports.DATE_FORMAT = exports.CALENDAR_FORMAT = void 0;
exports.areDatesOnSameDay = areDatesOnSameDay;
exports.getYearRange = exports.getWeekdaysShort = exports.getWeekdaysMin = exports.getWeekdays = exports.getToday = exports.getParsedDate = exports.getMonths = exports.getMonthName = exports.getMonthDays = exports.getFormatedDate = exports.getFormated = exports.getDateYear = exports.getDateMonth = exports.getDate = void 0;
var _dayjs = _interopRequireDefault(require("dayjs"));
function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }
const CALENDAR_FORMAT = exports.CALENDAR_FORMAT = 'YYYY-MM-DD HH:mm';
const DATE_FORMAT = exports.DATE_FORMAT = 'YYYY-MM-DD';
const YEAR_PAGE_SIZE = exports.YEAR_PAGE_SIZE = 12;
const getMonths = () => _dayjs.default.months();
exports.getMonths = getMonths;
const getMonthName = month => _dayjs.default.months()[month];
exports.getMonthName = getMonthName;
const getWeekdays = () => _dayjs.default.weekdays();
exports.getWeekdays = getWeekdays;
const getWeekdaysShort = () => _dayjs.default.weekdaysShort();
exports.getWeekdaysShort = getWeekdaysShort;
const getWeekdaysMin = () => _dayjs.default.weekdaysMin();
exports.getWeekdaysMin = getWeekdaysMin;
const getFormated = date => (0, _dayjs.default)(date).format(CALENDAR_FORMAT);
exports.getFormated = getFormated;
const getDateMonth = date => (0, _dayjs.default)(date).month();
exports.getDateMonth = getDateMonth;
const getDateYear = date => (0, _dayjs.default)(date).year();
exports.getDateYear = getDateYear;
const getToday = () => (0, _dayjs.default)().format(DATE_FORMAT);
exports.getToday = getToday;
function areDatesOnSameDay(a, b) {
  if (!a || !b) {
    return false;
  }
  const date_a = (0, _dayjs.default)(a).format(DATE_FORMAT);
  const date_b = (0, _dayjs.default)(b).format(DATE_FORMAT);
  return date_a === date_b;
}
const getFormatedDate = (date, format) => (0, _dayjs.default)(date).format(format);
exports.getFormatedDate = getFormatedDate;
const getDate = date => (0, _dayjs.default)(date, CALENDAR_FORMAT);
exports.getDate = getDate;
const getYearRange = year => {
  const endYear = YEAR_PAGE_SIZE * Math.ceil(year / YEAR_PAGE_SIZE);
  let startYear = endYear === year ? endYear : endYear - YEAR_PAGE_SIZE;
  if (startYear < 0) startYear = 0;
  return Array.from({
    length: YEAR_PAGE_SIZE
  }, (_, i) => startYear + i);
};

/**
 * Get detailed date object
 *
 * @param date Get detailed date object
 *
 * @returns parsed date object
 */
exports.getYearRange = getYearRange;
const getParsedDate = date => {
  return {
    year: (0, _dayjs.default)(date).year(),
    month: (0, _dayjs.default)(date).month(),
    hour: (0, _dayjs.default)(date).hour(),
    minute: (0, _dayjs.default)(date).minute()
  };
};

/**
 * Calculate month days array based on current date
 *
 * @param datetime - The current date that selected
 * @param displayFullDays
 * @param minimumDate - min selectable date
 * @param maximumDate - max selectable date
 *
 * @returns days array based on current date
 */
exports.getParsedDate = getParsedDate;
const getMonthDays = (datetime = (0, _dayjs.default)(), displayFullDays, minimumDate, maximumDate) => {
  const date = getDate(datetime);
  const daysInMonth = date.daysInMonth();
  const prevMonthDays = date.add(-1, 'month').daysInMonth();
  const firstDay = date.date(1);
  const dayOfMonth = firstDay.day() % 7;
  const prevDays = displayFullDays ? Array.from({
    length: dayOfMonth
  }, (_, i) => {
    const day = i + (prevMonthDays - dayOfMonth + 1);
    const thisDay = date.add(-1, 'month').date(day);
    return generateDayObject(day, thisDay, minimumDate, maximumDate, false);
  }) : Array(dayOfMonth).fill(null);
  const monthDaysOffset = dayOfMonth + daysInMonth;
  const nextMonthDays = displayFullDays ? monthDaysOffset > 35 ? 42 - monthDaysOffset : 35 - monthDaysOffset : 0;
  const currentDays = Array.from({
    length: daysInMonth
  }, (_, i) => {
    const day = i + 1;
    const thisDay = date.date(day);
    return generateDayObject(day, thisDay, minimumDate, maximumDate, true);
  });
  const nextDays = Array.from({
    length: nextMonthDays
  }, (_, i) => {
    const day = i + 1;
    const thisDay = date.add(1, 'month').date(day);
    return generateDayObject(day, thisDay, minimumDate, maximumDate, false);
  });
  return [...prevDays, ...currentDays, ...nextDays];
};

/**
 * Generate day object for displaying inside day cell
 *
 * @param day - number of day
 * @param date - calculated date based on day, month, and year
 * @param minDate - min selectable date
 * @param maxDate - max selectable date
 * @param isCurrentMonth - define the day is in the current month
 *
 * @returns days object based on current date
 */
exports.getMonthDays = getMonthDays;
const generateDayObject = (day, date, minDate, maxDate, isCurrentMonth) => {
  let disabled = false;
  if (minDate) {
    disabled = date < getDate(minDate);
  }
  if (maxDate && !disabled) {
    disabled = date > getDate(maxDate);
  }
  return {
    text: day.toString(),
    day: day,
    date: getFormatedDate(date, DATE_FORMAT),
    disabled,
    isCurrentMonth
  };
};
